contracts/PeeranhaCommunityTokenFactory.sol

```diff
@@ -0,0 +1,174 @@
+    IPeeranhaCommunity peeranhaCommunity;
+  }
+  FactoryData factoryData;
+
+  event CommunityTokenCreated(address indexed communityTokenContractAddress, uint32 indexed communityId);
+  event SetCommunityTokenPool(uint16 indexed period); 
+  event CommunityRewardSettingsUpdated(address indexed communityTokenContractAddress);
+  event PayCommunityRewards(address indexed userAddress, uint16 indexed period);
+
+  function initialize(address peeranhaUserContractAddress, address peeranhaCommunityContractAddress) public initializer {
+    factoryData.peeranhaUser = IPeeranhaUser(peeranhaUserContractAddress);
+    factoryData.peeranhaCommunity = IPeeranhaCommunity(peeranhaCommunityContractAddress);
+  }
+
+  function dispatcherCheck(address userAddress) internal {
+    if (userAddress != _msgSender()) {
+      factoryData.peeranhaUser.onlyDispatcher(_msgSender());
+    }
+  }
+
+  function createNewCommunityToken(address userAddress, uint32 communityId, address tokenAddress, uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod) external override {
+    dispatcherCheck(userAddress);
+    factoryData.peeranhaCommunity.onlyExistingAndNotFrozenCommunity(userAddress, communityId);
+    require(factoryData.peeranhaUser.isProtocolAdmin(userAddress), "not_allowed_not_protocal_admin");  // tests
+    factoryData.peeranhaCommunitiesToken[communityId].push(new PeeranhaCommunityToken(tokenAddress, maxRewardPerPeriod, activeUsersInPeriod, address(this)));
+    
+    bool isAddedCommunityId;
+    for (uint256 i; i < factoryData.factoryCommunitiesId.length; i++) {
+      if (factoryData.factoryCommunitiesId[i] == communityId)
+        isAddedCommunityId = true;
+    }
+    if (!isAddedCommunityId) {
+      factoryData.factoryCommunitiesId.push(communityId);
+    }
+
+    uint256 contractsCommunityTokenLength = factoryData.peeranhaCommunitiesToken[communityId].length;
+    address communityTokenAddress = address(factoryData.peeranhaCommunitiesToken[communityId][contractsCommunityTokenLength - 1]);
+    emit CommunityTokenCreated(communityTokenAddress, communityId);
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override
+      view
+      returns (address sender)
+  {
+      return NativeMetaTransaction._msgSender();
+  }
+
+  function updateCommunityRewardSettings(address userAddress, uint32 communityId, address communityTokenContractAddress, uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod) external override {
+    dispatcherCheck(userAddress);
+    IPeeranhaCommunityToken peeranhaCommunityToken = getContractCommunityToken(communityId, communityTokenContractAddress);
+    peeranhaCommunityToken.updateCommunityRewardSettings(maxRewardPerPeriod, activeUsersInPeriod);
+
+    emit CommunityRewardSettingsUpdated(communityTokenContractAddress);
+  }
+  
+  // set pools
+  function setTotalPeriodRewards(uint16 period) external override {
```

Comment:
Is it the action that we need to execute at the end of each period?
can we use `SetCommunityTokenPool` event to init periods data?